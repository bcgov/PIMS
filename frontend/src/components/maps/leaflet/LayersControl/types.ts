import { WMSOptions } from 'leaflet';
import { TreeNodeInArray } from 'react-simple-tree-menu';

export interface ILayerItem extends TreeNodeInArray, WMSOptions {}
