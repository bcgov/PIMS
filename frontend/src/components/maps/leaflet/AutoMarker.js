import { Marker } from 'react-leaflet';
import React from 'react';

// Create your own class, extending from the Marker class.
const AutoMarker = props => {

  const openPopup = marker => {
    if (marker && props.autoPopup) {
        marker.leafletElement.openPopup();
    }
  };

  return (
    <Marker ref={el => openPopup(el)} {...props}/>
  );
};
export default AutoMarker;