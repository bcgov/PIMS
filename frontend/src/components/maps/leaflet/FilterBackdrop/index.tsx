import * as React from 'react';
import { Spinner } from 'react-bootstrap';
import styled from 'styled-components';

export interface FilterBackdropProps {
  show?: boolean;
}

const Backdrop = styled.div`
  width: 100%;
  height: 100%;
  position: fixed;
  z-index: 999;
  left: 0;
  background-color: rgba(0, 0, 0, 0.4);
  display: flex;
  align-items: center;
  align-content: center;
  justify-items: center;
  justify-content: center;
`;

const FilterBackdrop: React.FC<FilterBackdropProps> = ({ show }) => {
  return show ? (
    <Backdrop>
      <Spinner animation="border" variant="warning" />
    </Backdrop>
  ) : null;
};

export default FilterBackdrop;
