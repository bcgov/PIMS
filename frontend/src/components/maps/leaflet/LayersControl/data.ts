import { ILayerItem } from './types';

export const layersTree: ILayerItem[] = [
  {
    key: 'Administrative Boundaries',
    label: 'Administrative Boundaries',
    on: false,
    nodes: [
      {
        key: 'currentEconomicRegions',
        label: 'Current Census Economic Regions',
        on: false,
        url: 'https://openmaps.gov.bc.ca/geo/pub/WHSE_HUMAN_CULTURAL_ECONOMIC.CEN_ECONOMIC_REGIONS_SVW/ows?',
        layers: 'pub:WHSE_HUMAN_CULTURAL_ECONOMIC.CEN_ECONOMIC_REGIONS_SVW',
        transparent: true,
        format: 'image/png',
        zIndex: 0,
        id: 'currentEconomicRegions',
        color: '#bbdefb',
        outline: '#7E8185',
      },
      {
        key: 'moti',
        label: 'MOTI Regional Boundaries',
        on: false,
        url: 'https://openmaps.gov.bc.ca/geo/pub/WHSE_ADMIN_BOUNDARIES.TADM_MOT_REGIONAL_BNDRY_POLY/ows?',
        layers: 'pub:WHSE_ADMIN_BOUNDARIES.TADM_MOT_REGIONAL_BNDRY_POLY',
        transparent: true,
        format: 'image/png',
        zIndex: 21,
        id: 'moti',
        color: ['#E4D597', '#F4EF75', '#E6C46D'],
        outline: '#ACA08B',
      },
      {
        key: 'municipalities',
        label: 'Municipalities',
        on: false,
        layers: 'pub:WHSE_LEGAL_ADMIN_BOUNDARIES.ABMS_MUNICIPALITIES_SP',
        url: 'https://openmaps.gov.bc.ca/geo/pub/WHSE_LEGAL_ADMIN_BOUNDARIES.ABMS_MUNICIPALITIES_SP/ows?',
        transparent: true,
        format: 'image/png',
        zIndex: 20,
        id: 'municipalityLayer',
        opacity: 0.5,
        color: '#f2b3ff',
        outline: '#747374',
      },
      {
        key: 'regionalDistricts ',
        label: 'Regional Districts ',
        on: false,
        layers: 'pub:WHSE_LEGAL_ADMIN_BOUNDARIES.ABMS_REGIONAL_DISTRICTS_SP',
        url: 'https://openmaps.gov.bc.ca/geo/pub/WHSE_LEGAL_ADMIN_BOUNDARIES.ABMS_REGIONAL_DISTRICTS_SP/ows?',
        transparent: true,
        format: 'image/png',
        zIndex: 19,
        id: 'regionalDistricts',
        opacity: 0.5,
      },
    ],
  },
  {
    key: 'firstNations',
    label: 'First Nations',
    on: false,
    nodes: [
      {
        key: 'firstNationsReserves',
        label: 'First Nations reserves',
        on: false,
        url: 'https://openmaps.gov.bc.ca/geo/pub/WHSE_ADMIN_BOUNDARIES.ADM_INDIAN_RESERVES_BANDS_SP/ows?',
        layers: 'pub:WHSE_ADMIN_BOUNDARIES.ADM_INDIAN_RESERVES_BANDS_SP',
        transparent: true,
        format: 'image/png',
        zIndex: 10,
        id: 'firstNationsReserves',
        color: '#ebe0d7',
        outline: '#8D847F',
      },
      {
        key: 'firstNationTreatyAreas',
        label: 'First Nation Treaty Areas',
        on: false,
        url: 'https://openmaps.gov.bc.ca/geo/pub/WHSE_LEGAL_ADMIN_BOUNDARIES.FNT_TREATY_AREA_SP/ows?',
        layers: 'pub:WHSE_LEGAL_ADMIN_BOUNDARIES.FNT_TREATY_AREA_SP',
        transparent: true,
        format: 'image/png',
        zIndex: 10,
        id: 'firstNationTreatyAreas',
      },
      {
        key: 'firstNationTreatyLands',
        label: 'First Nations Treaty Lands',
        on: false,
        url: 'https://openmaps.gov.bc.ca/geo/pub/WHSE_LEGAL_ADMIN_BOUNDARIES.FNT_TREATY_LAND_SP/ows?',
        layers: 'pub:WHSE_LEGAL_ADMIN_BOUNDARIES.FNT_TREATY_LAND_SP',
        transparent: true,
        format: 'image/png',
        zIndex: 10,
        id: 'firstNationTreatyLands',
      },
      {
        key: 'firstNationTreatyRelatedLands',
        label: 'First Nations Treaty Related Lands',
        on: false,
        url: 'https://openmaps.gov.bc.ca/geo/pub/WHSE_LEGAL_ADMIN_BOUNDARIES.FNT_TREATY_RELATED_LAND_SP/ows?',
        layers: 'pub:WHSE_LEGAL_ADMIN_BOUNDARIES.FNT_TREATY_RELATED_LAND_SP',
        transparent: true,
        format: 'image/png',
        zIndex: 10,
        id: 'firstNationTreatyRelatedLands',
      },
      {
        key: 'firstNationTreatySideAgreement',
        label: 'First Nation Treaty Side Agreements',
        on: false,
        url: 'https://openmaps.gov.bc.ca/geo/pub/WHSE_LEGAL_ADMIN_BOUNDARIES.FNT_TREATY_SIDE_AGREEMENTS_SP/ows?',
        layers: 'pub:WHSE_LEGAL_ADMIN_BOUNDARIES.FNT_TREATY_SIDE_AGREEMENTS_SP',
        transparent: true,
        format: 'image/png',
        zIndex: 10,
        id: 'firstNationTreatySideAgreement',
        color: '#F8D5D2',
        outline: '#D18D84',
      },
    ],
  },
  {
    key: 'landOwnership',
    label: 'Land Ownership',
    on: false,
    nodes: [
      {
        key: 'crownLeases',
        label: 'Crown Leases',
        on: false,
        layers: 'pub:WHSE_TANTALIS.TA_CROWN_LEASES_SVW',
        url: 'https://openmaps.gov.bc.ca/geo/pub/WHSE_TANTALIS.TA_CROWN_LEASES_SVW/ows?',
        transparent: true,
        format: 'image/png',
        zIndex: 7,
        id: 'crownLeases',
        opacity: 0.5,
        color: 'transparent',
        outline: '#1F6182',
        stripes: '#1F6182',
      },
      {
        key: 'crownInventory',
        label: 'Crown Inventory',
        on: false,
        layers: 'pub:WHSE_TANTALIS.TA_CROWN_INVENTORY_SVW',
        url: 'https://openmaps.gov.bc.ca/geo/pub/WHSE_TANTALIS.TA_CROWN_INVENTORY_SVW/ows?',
        transparent: true,
        format: 'image/png',
        zIndex: 6,
        id: 'crownInventory',
        opacity: 0.5,
        color: '#fcc1eb',
        outline: '#A286A1',
      },
      {
        key: 'crownLandLicenses',
        label: 'Crown Land Licenses',
        on: false,
        layers: 'pub:WHSE_TANTALIS.TA_CROWN_LICENSES_SVW',
        url: 'https://openmaps.gov.bc.ca/geo/pub/WHSE_TANTALIS.TA_CROWN_LICENSES_SVW/ows?',
        transparent: true,
        format: 'image/png',
        zIndex: 5,
        id: 'crownLandLicenses',
        opacity: 0.5,
        color: 'transparent',
        outline: '#12387E',
        stripes: '#12387E',
      },
      {
        key: 'parcelBoundaries',
        label: 'Parcel Boundaries',
        on: true,
        url: 'https://openmaps.gov.bc.ca/geo/pub/WHSE_CADASTRE.PMBC_PARCEL_FABRIC_POLY_SVW/ows?',
        layers: 'pub:WHSE_CADASTRE.PMBC_PARCEL_FABRIC_POLY_SVW',
        transparent: true,
        format: 'image/png',
        zIndex: 18,
        id: 'parcelLayer',
        color: 'transparent',
        outline: '#E6A018',
      },
    ],
  },
  {
    key: 'zoning',
    label: 'Zoning',
    on: false,
    nodes: [
      {
        key: 'agriculturalLandReserveLines',
        label: 'Agricultural Land Reserve Lines',
        on: false,
        layers: 'pub:WHSE_LEGAL_ADMIN_BOUNDARIES.OATS_ALR_BOUNDARY_LINES_SVW',
        url: 'https://openmaps.gov.bc.ca/geo/pub/WHSE_LEGAL_ADMIN_BOUNDARIES.OATS_ALR_BOUNDARY_LINES_SVW/ows?',
        transparent: true,
        format: 'image/png',
        zIndex: 4,
        id: 'agriculturalLandReserveLines',
        opacity: 0.5,
        color: 'transparent',
        outline: '#00b300',
      },
    ],
  },
  {
    key: 'disturbances',
    label: 'Disturbances',
    on: false,
    nodes: [
      {
        key: 'environmentalRemediationSites',
        label: 'Environmental Remediation Sites',
        on: false,
        layers: 'pub:WHSE_WASTE.SITE_ENV_RMDTN_SITES_SVW',
        url: 'https://openmaps.gov.bc.ca/geo/pub/WHSE_WASTE.SITE_ENV_RMDTN_SITES_SVW/ows?',
        transparent: true,
        format: 'image/png',
        zIndex: 4,
        id: 'environmentalRemediationSites',
        opacity: 0.5,
        color: '#DC0000',
        outline: 'black',
        circle: true,
      },
    ],
  },
];
