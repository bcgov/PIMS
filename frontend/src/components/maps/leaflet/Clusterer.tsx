import React, { PropsWithChildren } from 'react';
import { LatLng, LeafletMouseEvent, Icon, DivIcon, Layer, Marker, GeoJSONEvent } from 'leaflet';
import { Map as LeafletMap, TileLayer, Popup, WMSTileLayer, GeoJSON } from 'react-leaflet';
import { BBox, GeoJsonProperties } from 'geojson';
import Supercluster from 'supercluster';
import useSupercluster, { ICluster } from 'hooks/useSupercluster';
import { createSingleMarker, createClusterMarker } from './mapUtils';
import { uniqueId } from 'lodash';

export type ClustererProps<P> = {
  points: Supercluster.PointFeature<P>[];
  bounds?: BBox;
  zoom: number;
  onMarkerClick?: Function;
  onClusterClick?: Function;
};

const Clusterer = <P extends GeoJsonProperties>(props: PropsWithChildren<ClustererProps<P>>) => {
  // state and refs
  const { points, bounds, zoom, onMarkerClick, onClusterClick } = props;

  // get clusters
  // clusters are an array of GeoJSON Feature objects, but some of them
  // represent a cluster of points, and some represent individual points.
  const { clusters, supercluster } = useSupercluster({
    points,
    bounds,
    zoom,
    options: { radius: 75, maxZoom: 18 },
  });

  const pointToLayer = (feature: ICluster, latlng: LatLng): Layer => {
    const { cluster: isCluster } = feature?.properties;
    // we have a cluster to render
    if (!!isCluster) {
      return createClusterMarker(feature, latlng);
    }
    // we have a single point (parcel or building) to render
    return createSingleMarker(feature, latlng);
  };

  const onClick = (e: LeafletMouseEvent) => {
    // the point may be either a cluster or a single map pin
    const cluster = (e?.layer as Marker)?.feature as ICluster;
    const isCluster = cluster?.properties?.cluster;

    // the user clicked on a cluster
    if (isCluster) {
      onClusterClick?.(cluster);
    } else {
      // the user clicked on a single map pin
      onMarkerClick?.(cluster);
    }
  };

  // TODO: improve typing
  return clusters && clusters?.length > 0 ? (
    <GeoJSON
      key={uniqueId('geojson_')}
      data={clusters as any}
      pointToLayer={pointToLayer}
      onclick={onClick}
    ></GeoJSON>
  ) : null;
};

export default Clusterer;
