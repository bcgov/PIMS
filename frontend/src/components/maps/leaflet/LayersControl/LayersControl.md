LayerControl example:

```js
import 'leaflet/dist/leaflet.css';
import '../Map.scss';
import 'leaflet';
import { TileLayer, Map as MapContainer } from 'react-leaflet';

<MapContainer center={[48.423078, -123.360956]} zoom={18}>
  <TileLayer
    attribution='&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors'
    url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
  />

  <LayersControl />
</MapContainer>;
```
