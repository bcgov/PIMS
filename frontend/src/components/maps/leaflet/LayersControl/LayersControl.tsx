import variables from '_variables.module.scss';
import clsx from 'classnames';
import TooltipWrapper from 'components/common/TooltipWrapper';
import * as L from 'leaflet';
import * as React from 'react';
import { useEffect } from 'react';
import { Button } from 'react-bootstrap';
import { FaLayerGroup } from 'react-icons/fa';
import Control from 'react-leaflet-control';
import styled from 'styled-components';

import LayersTree from './LayersTree';

const LayersContainer = styled.div`
  margin-right: -10px;
  width: 341px;
  min-height: 52px;
  height: 500px;
  max-height: 500px;
  background-color: #fff;
  position: relative;
  border-radius: 4px;
  box-shadow: -2px 1px 4px rgba(0, 0, 0, 0.2);
  z-index: 1000;

  &.closed {
    width: 0px;
    height: 0px;
  }
`;

const LayersHeader = styled.div`
  width: 100%;
  height: 80px;
  background-color: ${variables.slideOutBlue};
  color: #fff;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-evenly;
  padding-top: 16px;
  border-top-right-radius: 4px;
  border-top-left-radius: 4px;
`;

const LayersContent = styled.div`
  width: 100%;
  height: calc(100% - 80px);
  padding: 16px;

  &.open {
    overflow-y: scroll;
  }
`;

const LayersIcon = styled(FaLayerGroup)`
  font-size: 30px;
`;

const Title = styled.p`
  font-size: 18px;
  color: #ffffff;
  text-decoration: none solid rgb(255, 255, 255);
  line-height: 18px;
  font-weight: bold;
`;

const ControlButton = styled(Button)`
  width: 52px;
  height: 52px;
  position: absolute;
  top: 0;
  left: -51px;
  background-color: #fff;
  color: ${variables.slideOutBlue};
  border-color: ${variables.slideOutBlue};
  box-shadow: -2px 1px 4px rgba(0, 0, 0, 0.2);
  &.open {
    border-top-right-radius: 0;
    border-bottom-right-radius: 0;
    top: 26px;
  }
`;

export type ILayersControl = {
  /** whether the slide out is open or closed */
  open: boolean;
  /** set the slide out as open or closed */
  setOpen: () => void;
};

/**
 * Component to display the layers control on the map
 * @example ./LayersControl.md
 */
const LayersControl: React.FC<ILayersControl> = ({ open, setOpen }) => {
  useEffect(() => {
    const elem = L.DomUtil.get('layersContainer');
    if (elem) {
      L.DomEvent.on(elem!, 'mousewheel', L.DomEvent.stopPropagation);
    }
  });

  return (
    <Control position="topright">
      <LayersContainer id="layersContainer" className={clsx({ closed: !open })}>
        {open && (
          <LayersHeader>
            <LayersIcon />
            <Title>Layers</Title>
          </LayersHeader>
        )}
        <TooltipWrapper toolTipId="layer-control-id" toolTip="Layer Controls">
          <ControlButton
            id="layersControlButton"
            variant="outline-secondary"
            onClick={setOpen}
            className={clsx({ open })}
          >
            <LayersIcon />
          </ControlButton>
        </TooltipWrapper>
        <LayersContent className={clsx({ open })}>
          <LayersTree />
        </LayersContent>
      </LayersContainer>
    </Control>
  );
};

export default LayersControl;
