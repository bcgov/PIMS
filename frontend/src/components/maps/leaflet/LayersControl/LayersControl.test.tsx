import 'leaflet';
import 'leaflet/dist/leaflet.css';

import { cleanup, render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import L from 'leaflet';
import { noop } from 'lodash';
import { useState } from 'react';
import React from 'react';
import { createMapContainer, deferred } from 'utils/testUtils';

import LayersControl from './LayersControl';

jest.mock('axios');

// component under test
function Template({ openByDefault = false }) {
  const [open, setOpen] = useState(openByDefault);
  return <LayersControl open={open} setOpen={() => setOpen(!open)} />;
}

function setup(ui = <Template />, setMap = noop) {
  // create a promise to wait for the map to be ready (which happens after initial render)
  const { promise, resolve } = deferred();
  const wrapper = createMapContainer(resolve, setMap);
  const component = render(ui, { wrapper });
  return {
    component,
    ready: promise,
    findLayerList: () => document.querySelector('#layersContainer') as HTMLElement,
    findToggleButton: () => document.querySelector('#layersControlButton') as HTMLElement,
  };
}

function isLayerVisible(key: string, leaflet: any) {
  return Object.keys(leaflet._layers)
    .map((k) => leaflet._layers[k])
    .map((x) => x.options)
    .find((options) => options?.key === key);
}

describe('LayersControl View', () => {
  beforeAll(() => {
    (global as any).IS_REACT_ACT_ENVIRONMENT = false;
  });
  afterEach(cleanup);

  it('renders correctly', () => {
    const { component } = setup();
    expect(component.asFragment()).toMatchSnapshot();
  });

  it('should render the layers control button', async () => {
    const { ready, findToggleButton } = setup();
    await waitFor(() => ready);
    const toggleBtn = findToggleButton();
    expect(toggleBtn).toBeInTheDocument();
  });

  it('should be closed by default', async () => {
    const { ready, findLayerList } = setup();
    await waitFor(() => ready);
    const layersContainer = findLayerList();
    expect(layersContainer).toBeInTheDocument();
    expect(layersContainer.className).toContain('closed');
  });

  it('when closed, clicking the button should open the layer list', async () => {
    const { ready, findLayerList, findToggleButton } = setup();
    await waitFor(() => ready);
    // when layer list is closed...
    const layersContainer = findLayerList();
    expect(layersContainer).toBeInTheDocument();
    expect(layersContainer.className).toContain('closed');
    // clicking the button should open it...
    const toggleBtn = findToggleButton();
    userEvent.click(toggleBtn);
    await waitFor(() => expect(layersContainer.className).not.toContain('closed'));
  });

  it('when open, clicking the button should close the layers list', async () => {
    const { ready, findLayerList, findToggleButton } = setup(<Template openByDefault={true} />);
    await waitFor(() => ready);
    // when layer list is open...
    const layersContainer = findLayerList();
    expect(layersContainer).toBeInTheDocument();
    expect(layersContainer.className).not.toContain('closed');
    // clicking the button should close it...
    const toggleBtn = findToggleButton();
    userEvent.click(toggleBtn);
    await waitFor(() => expect(layersContainer.className).toContain('closed'));
  });

  it('should enable the Parcels layer and disable the Municipality layer by default', async () => {
    let mapInstance: L.Map | undefined = undefined;
    function setMap(map: L.Map) {
      mapInstance = map;
    }
    const { ready } = setup(<Template />, setMap);
    await waitFor(() => ready);
    await waitFor(() => expect(mapInstance).toBeDefined());

    expect(isLayerVisible('parcelBoundaries', mapInstance)).toBeDefined();
    expect(isLayerVisible('municipalities', mapInstance)).toBeUndefined();
  });
});
