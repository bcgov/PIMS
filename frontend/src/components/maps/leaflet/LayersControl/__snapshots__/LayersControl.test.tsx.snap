// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayersControl View renders correctly 1`] = `
<DocumentFragment>
  .c4 {
  display: flex;
  align-items: center;
  padding-left: 0px;
  border: none;
  padding-top: 5px;
  padding-bottom: 5px;
}

.c4 .form-group .form-check label {
  font-weight: bold;
  color: textColor;
  font-size: 13px;
}

.c5 {
  margin-right: 10px;
  font-size: 15px;
}

.c6 {
  margin-bottom: 0px;
}

.c6 .form-check {
  display: flex;
  align-items: center;
}

.c6 .form-check input {
  margin-top: 0px;
}

.c6 .form-check label {
  display: flex;
  align-items: center;
}

.c0 {
  margin-right: -10px;
  width: 341px;
  min-height: 52px;
  height: 500px;
  max-height: 500px;
  background-color: #fff;
  position: relative;
  border-radius: 4px;
  box-shadow: -2px 1px 4px rgba(0, 0, 0, 0.2);
  z-index: 1000;
}

.c0.closed {
  width: 0px;
  height: 0px;
}

.c3 {
  width: 100%;
  height: calc(100% - 80px);
  padding: 16px;
}

.c3.open {
  overflow-y: scroll;
}

.c2 {
  font-size: 30px;
}

.c1 {
  width: 52px;
  height: 52px;
  position: absolute;
  top: 0;
  left: -51px;
  background-color: #fff;
  color: slideOutBlue;
  border-color: slideOutBlue;
  box-shadow: -2px 1px 4px rgba(0, 0, 0, 0.2);
}

.c1.open {
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
  top: 26px;
}

<div
    id="mapid"
    style="width: 500px; height: 500px;"
  >
    <div
      class="leaflet-container leaflet-touch leaflet-grab leaflet-touch-drag leaflet-touch-zoom"
      style="position: relative;"
      tabindex="0"
    >
      <div
        class="leaflet-pane leaflet-map-pane"
        style="left: 0px; top: 0px;"
      >
        <div
          class="leaflet-pane leaflet-tile-pane"
        >
          <div
            class="leaflet-layer "
            style="z-index: 9;"
          >
            <div
              class="leaflet-tile-container leaflet-zoom-animated"
              style="z-index: 18; left: 0px; top: 0px;"
            >
              <img
                alt=""
                class="leaflet-tile"
                src="https://openmaps.gov.bc.ca/geo/pub/WHSE_CADASTRE.PMBC_PARCEL_FABRIC_POLY_SVW/ows?&service=WMS&request=GetMap&layers=pub%3AWHSE_CADASTRE.PMBC_PARCEL_FABRIC_POLY_SVW&styles=&format=image%2Fpng&transparent=true&version=1.1.1&key=parcelBoundaries&label=Parcel%20Boundaries&on=true&url=https%3A%2F%2Fopenmaps.gov.bc.ca%2Fgeo%2Fpub%2FWHSE_CADASTRE.PMBC_PARCEL_FABRIC_POLY_SVW%2Fows%3F&id=parcelLayer&color=transparent&outline=%23E6A018&width=256&height=256&srs=EPSG%3A3857&bbox=-13734205.24228047,6178557.870347368,-13731759.257375345,6181003.855252495"
                style="width: 256px; height: 256px; left: -75px; top: -242px;"
              />
            </div>
          </div>
        </div>
        <div
          class="leaflet-pane leaflet-overlay-pane"
        />
        <div
          class="leaflet-pane leaflet-shadow-pane"
        />
        <div
          class="leaflet-pane leaflet-marker-pane"
        />
        <div
          class="leaflet-pane leaflet-tooltip-pane"
        />
        <div
          class="leaflet-pane leaflet-popup-pane"
        />
      </div>
      <div
        class="leaflet-control-container"
      >
        <div
          class="leaflet-top leaflet-left"
        >
          <div
            class="leaflet-control-zoom leaflet-bar leaflet-control"
          >
            <a
              aria-disabled="false"
              aria-label="Zoom in"
              class="leaflet-control-zoom-in"
              href="#"
              role="button"
              title="Zoom in"
            >
              <span
                aria-hidden="true"
              >
                +
              </span>
            </a>
            <a
              aria-disabled="false"
              aria-label="Zoom out"
              class="leaflet-control-zoom-out"
              href="#"
              role="button"
              title="Zoom out"
            >
              <span
                aria-hidden="true"
              >
                âˆ’
              </span>
            </a>
          </div>
        </div>
        <div
          class="leaflet-top leaflet-right"
        >
          <div
            class="leaflet-control"
          >
            <div
              class="c0 closed"
              id="layersContainer"
            >
              <div>
                <button
                  class="c1 btn btn-outline-secondary"
                  id="layersControlButton"
                  type="button"
                >
                  <svg
                    class="c2"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 512 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M12.41 148.02l232.94 105.67c6.8 3.09 14.49 3.09 21.29 0l232.94-105.67c16.55-7.51 16.55-32.52 0-40.03L266.65 2.31a25.607 25.607 0 0 0-21.29 0L12.41 107.98c-16.55 7.51-16.55 32.53 0 40.04zm487.18 88.28l-58.09-26.33-161.64 73.27c-7.56 3.43-15.59 5.17-23.86 5.17s-16.29-1.74-23.86-5.17L70.51 209.97l-58.1 26.33c-16.55 7.5-16.55 32.5 0 40l232.94 105.59c6.8 3.08 14.49 3.08 21.29 0L499.59 276.3c16.55-7.5 16.55-32.5 0-40zm0 127.8l-57.87-26.23-161.86 73.37c-7.56 3.43-15.59 5.17-23.86 5.17s-16.29-1.74-23.86-5.17L70.29 337.87 12.41 364.1c-16.55 7.5-16.55 32.5 0 40l232.94 105.59c6.8 3.08 14.49 3.08 21.29 0L499.59 404.1c16.55-7.5 16.55-32.5 0-40z"
                    />
                  </svg>
                </button>
              </div>
              <div
                class="c3"
              >
                <form
                  action="#"
                >
                  <div
                    tabindex="0"
                  >
                    <div
                      class="list-group"
                    >
                      <div
                        class="c4 list-group-item"
                        id="Administrative Boundaries"
                      >
                        <svg
                          class="c5"
                          fill="currentColor"
                          height="1em"
                          stroke="currentColor"
                          stroke-width="0"
                          viewBox="0 0 256 512"
                          width="1em"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                          />
                        </svg>
                        <div
                          class="c6"
                        >
                          <div
                            class="form-check"
                          >
                            <input
                              class="form-check-input"
                              type="checkbox"
                            />
                            <label
                              class="form-check-label"
                              title=""
                            >
                               Administrative Boundaries
                            </label>
                          </div>
                        </div>
                      </div>
                      <div
                        class="c4 list-group-item"
                        id="firstNations"
                      >
                        <svg
                          class="c5"
                          fill="currentColor"
                          height="1em"
                          stroke="currentColor"
                          stroke-width="0"
                          viewBox="0 0 256 512"
                          width="1em"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                          />
                        </svg>
                        <div
                          class="c6"
                        >
                          <div
                            class="form-check"
                          >
                            <input
                              class="form-check-input"
                              type="checkbox"
                            />
                            <label
                              class="form-check-label"
                              title=""
                            >
                               First Nations
                            </label>
                          </div>
                        </div>
                      </div>
                      <div
                        class="c4 list-group-item"
                        id="landOwnership"
                      >
                        <svg
                          class="c5"
                          fill="currentColor"
                          height="1em"
                          stroke="currentColor"
                          stroke-width="0"
                          viewBox="0 0 256 512"
                          width="1em"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                          />
                        </svg>
                        <div
                          class="c6"
                        >
                          <div
                            class="form-check"
                          >
                            <input
                              class="form-check-input"
                              type="checkbox"
                            />
                            <label
                              class="form-check-label"
                              title=""
                            >
                               Land Ownership
                            </label>
                          </div>
                        </div>
                      </div>
                      <div
                        class="c4 list-group-item"
                        id="zoning"
                      >
                        <svg
                          class="c5"
                          fill="currentColor"
                          height="1em"
                          stroke="currentColor"
                          stroke-width="0"
                          viewBox="0 0 256 512"
                          width="1em"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                          />
                        </svg>
                        <div
                          class="c6"
                        >
                          <div
                            class="form-check"
                          >
                            <input
                              class="form-check-input"
                              type="checkbox"
                            />
                            <label
                              class="form-check-label"
                              title=""
                            >
                               Zoning
                            </label>
                          </div>
                        </div>
                      </div>
                      <div
                        class="c4 list-group-item"
                        id="disturbances"
                      >
                        <svg
                          class="c5"
                          fill="currentColor"
                          height="1em"
                          stroke="currentColor"
                          stroke-width="0"
                          viewBox="0 0 256 512"
                          width="1em"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                          />
                        </svg>
                        <div
                          class="c6"
                        >
                          <div
                            class="form-check"
                          >
                            <input
                              class="form-check-input"
                              type="checkbox"
                            />
                            <label
                              class="form-check-label"
                              title=""
                            >
                               Disturbances
                            </label>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </form>
              </div>
            </div>
          </div>
        </div>
        <div
          class="leaflet-bottom leaflet-left"
        />
        <div
          class="leaflet-bottom leaflet-right"
        >
          <div
            class="leaflet-control-attribution leaflet-control"
          >
            <a
              href="https://leafletjs.com"
              title="A JavaScript library for interactive maps"
            >
              Leaflet
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
